src/ast2ram/ClauseTranslator.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2013, 2015, Oracle and/or its affiliates. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file ClauseTranslator.h
+ *
+ * Translator for clauses from AST to RAM
+ *
+ ***********************************************************************/
+
+#pragma once
+
+#include "ast2ram/AstToRamTranslator.h"
+#include "ast2ram/ValueIndex.h"
+#include <map>
+#include <vector>
+
+namespace souffle::ast {
+class Argument;
+class Clause;
+class Node;
+}  // namespace souffle::ast
+
+namespace souffle::ast::analysis {
+class AuxiliaryArityAnalsyis;
+}
+
+namespace souffle::ram {
+class Operation;
+class Condition;
+class RelationReference;
+}  // namespace souffle::ram
+
+namespace souffle::ast2ram {
+
+class ClauseTranslator {
+public:
+    ClauseTranslator(AstToRamTranslator& translator)
+            : translator(translator), auxArityAnalysis(translator.getAuxArityAnalysis()) {}
+
+    Own<ram::Statement> translateClause(
+            const ast::Clause& clause, const ast::Clause& originalClause, const int version = 0);
+
+protected:
+    AstToRamTranslator& translator;
+
+    // create value index
+    ValueIndex valueIndex;
+
+    // current nesting level
+    int level = 0;
+
+    virtual Own<ram::Operation> createOperation(const ast::Clause& clause);
+    virtual Own<ram::Condition> createCondition(const ast::Clause& originalClause);
```

Comment:
Not at the moment, because the main clause translation requires the clause head, and the overloaded method in the provenance clause translation requires the body. This may change when provenance is refactored.